pkg/extensions/tbtc/tbtc_test.go

```diff
@@ -0,0 +1,428 @@
+	tbtc := &tbtc{tbtcChain}
+
+	err := tbtc.monitorRetrievePubKey(
+		ctx,
+		constantBackoff,
+		timeout,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tbtcChain.CreateDeposit(depositAddress)
+
+	keepPubkey, err := submitKeepPublicKey(depositAddress, tbtcChain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait a while before triggering the stop event because the
+	// extension must have time to handle the start event
+	time.Sleep(100 * time.Millisecond)
+
+	err = tbtcChain.RetrieveSignerPubkey(depositAddress)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait a bit longer than the monitoring timeout
+	// to make sure the potential transaction completes
+	time.Sleep(2 * timeout)
+
+	expectedRetrieveSignerPubkeyCalls := 1
+	actualRetrieveSignerPubkeyCalls := tbtcChain.Logger().RetrieveSignerPubkeyCalls()
+	if expectedRetrieveSignerPubkeyCalls != actualRetrieveSignerPubkeyCalls {
+		t.Errorf(
+			"unexpected number of RetrieveSignerPubkey calls\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			expectedRetrieveSignerPubkeyCalls,
+			actualRetrieveSignerPubkeyCalls,
+		)
+	}
+
+	depositPubkey, err := tbtcChain.DepositPubkey(depositAddress)
+	if err != nil {
+		t.Errorf("unexpected error while fetching deposit pubkey: [%v]", err)
+	}
+
+	if !bytes.Equal(keepPubkey[:], depositPubkey) {
+		t.Errorf(
+			"unexpected public key\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			keepPubkey,
+			depositPubkey,
+		)
+	}
+}
+
+func TestRetrievePubkey_KeepClosedEventOccurred(t *testing.T) {
+	ctx := context.Background()
+	tbtcChain := local.NewTBTCLocalChain()
+	tbtc := &tbtc{tbtcChain}
+
+	err := tbtc.monitorRetrievePubKey(
+		ctx,
+		constantBackoff,
+		timeout,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tbtcChain.CreateDeposit(depositAddress)
+
+	_, err = submitKeepPublicKey(depositAddress, tbtcChain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait a while before triggering the keep closed event because the
+	// extension must have time to handle the start event
+	time.Sleep(100 * time.Millisecond)
+
+	err = closeKeep(depositAddress, tbtcChain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait a bit longer than the monitoring timeout
+	// to make sure the potential transaction completes
+	time.Sleep(2 * timeout)
+
+	expectedRetrieveSignerPubkeyCalls := 0
+	actualRetrieveSignerPubkeyCalls := tbtcChain.Logger().RetrieveSignerPubkeyCalls()
+	if expectedRetrieveSignerPubkeyCalls != actualRetrieveSignerPubkeyCalls {
+		t.Errorf(
+			"unexpected number of RetrieveSignerPubkey calls\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			expectedRetrieveSignerPubkeyCalls,
+			actualRetrieveSignerPubkeyCalls,
+		)
+	}
+
+	_, err = tbtcChain.DepositPubkey(depositAddress)
+
+	expectedError := fmt.Errorf("no pubkey for deposit [%v]", depositAddress)
+	if !reflect.DeepEqual(expectedError, err) {
+		t.Errorf(
+			"unexpected error\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			expectedError,
+			err,
+		)
+	}
```

Comment:
Should we check there is no `DepositPubkey` if we already asserted `RetrieveSigner` was not called? It looks to me that we are testing the logic of `local` chain implementation this way 🤔

Reply:
I think having this check shows better what outcomes are we expecting in real world. I see that as follows:
1. Keep has been closed
2. The monitoring of retrieve pubkey has been stopped in result
3. Because the monitoring has been stopped, `RetrieveSignerPubkey` was not called
4. Because `RetrieveSignerPubkey` was not called, the deposit pubkey is not set

Reply:
Yeah, I see your point. From the perspective of defining the full expectations of how the system should behave it's indeed better.